targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F429xI/TARGET_NUCLEO_F429ZI/PinNames.h

```diff
@@ -38,6 +38,9 @@
 extern "C" {
 #endif
 
+/* If this macro is defined, then constexpr utility functions for pin-map seach can be used. */
+#define EXPLICIT_PINMAP_READY
```

Comment:
I'm thinking you might need a define per peripheral - this currently means "I have tables for all things there are xxx_init_direct", but isn't that set open to future expansion?

Also, most of our HAL options are explicitly defined to `0` or `1`, and tested with `#if` rather than `#ifdef` - I suggest following that model.

Reply:
I can't see the reason for adding define per peripheral. What would be a use case for this?

We can have a target which does not have SPI for example.  So what is missing is that 
 `get_spi_pinmap()` should be guarded with `#if DEVICE_SPI` (and other utility functions accordingly).

Reply:
I think my concern was mainly about what if you add more device types to this infrastructure? Is it the complete set now?

I guess it's unnecessary if that define commits that HAL to be upgraded whenever any new device type comes along.

Reply:
What do you mean by a `device types`? Do you mean device which has SPI, I2C, etc in different configurations?

Reply:
I meant some device outside the set you've implemented here (PWM, AnalogIn,  AnalogOut, SPI, 
I2C, Serial). If those are all the needs-pin-mapping devices then there's no issue.

Reply:
CAN is available also on STM boards.

Reply:
Right. So if you create `get_can_pinmap` at some point, then you'd have to add it to all `EXPLICIT_PINMAP_READY && DEVICE_CAN` boards immediately. But that's fine.

Reply:
I have investigated this problem and the problem exists 😢 
Lets consider two boards:
- K64F (explicit pin-map supported => `EXPLICIT_PINMAP_READY == 1`)
- K66F (explicit pin-map not supported => `EXPLICIT_PINMAP_READY == 0`).

Now everything is ok if different drivers are used for both targets like SPI. But for example AnalogIn driver is common for both targets, so now `analogin_init_direct()` is implemented for `K66F` and `EXPLICIT_PINMAP_READY == 0`. This is wrong. If I implement default `xxx_init_direct()` functions to call `init_direct()` as you suggested, then I will have a problem in the above case since `get_analogin_pinmap()` will return `{pin, NC, NC}` and this will be passed to valid `analogin_init_direct()`.

Reply:
Right, so that's about the 2 levels of implementation - "do I have a HAL direct init call?" (certain boards, unflagged - handled via "weak") and "can I do compile-time mapping?" (EXPLICIT_PINMAP_READY).

If we want to permit the intermediate level, then we need the direct call to be real, so it can't be the result from a dummy mapping call.

I think at this point we want to ask the question - do we really want to support the intermediate form? It was what you were originally planning, but that was with the broad all-platform approach. If you're targetting golden boards, how about going straight to the full implementation?

Reply:
(So resolve this specific problem by pushing K66F one way or the other - remove its `analogin_init_direct` (make it `#if EXPLICIT_PINMAP_READY`), or implement the tables for it.)

Reply:
Yes, I was thinking that the best way would be to make `analogin_init_direct()` `#if EXPLICIT_PINMAP_READY`.

Reply:
I would be fine with that, but it's a global decision. If you're going to attempt to simplify out those `#ifdef`s and have the 3rd constexpr form work on all platforms, via the "working default weak implementation" technique, then all targets must have their `init_direct` conditional on `EXPLICIT_PINMAP_READY`.

That effectively rules out the 2nd "target-specific manual configuration" form you originally envisaged from ever being used. All platforms that would benefit would actually have the `get_pinmap`s that rendered it redundant.

Reply:
(As a corollary, for cross-checking, that then means the default implementations DON'T have to be weak. They can just be conditional on `!EXPLICIT_PINMAP_READY`. Then if any "non-ready" target accidentally does define it, compile error.)

Reply:
We can't also go this way. Because new `xxx_init()` functions call `xxx_init_direct()`.

Reply:
Good point. That complicates it.

But that could be dealt with internally, for those targets with common code and not always ready:

    static void _spi_init_direct(const spi_pinmap_t *map)
    {
        ...
    }

    void spi_init(PinName, ...)
    {
        ...
        _spi_init_direct(&map);
    }
    
    #if EXPLICIT_PINMAP_READY
    void spi_init_direct(const spi_pinmap_t *map)
    {
        _spi_init_direct(map);
    }
    #endif

Reply:
This is getting too complicated. I suggest to take a step back and indicate failure if `xxx_init_direct()` is not implemented (weak implementation is provided). And define `xxx_init_direct()` under `#if EXPLICIT_PINMAP_READY`.